application/src/main/java/org/thingsboard/server/service/cloud/KafkaCloudEventService.java

```diff
@@ -0,0 +1,194 @@
+
+        EdgeEventMsgProto cloudEventMsgProto = ProtoUtils.toProto(cloudEvent);
+        ToCloudEventMsg toCloudEventMsg = ToCloudEventMsg.newBuilder().setCloudEventMsg(cloudEventMsgProto).build();
+
+        UUID entityId = cloudEvent.getEntityId() == null ? UUID.fromString(cloudEvent.getEntityBody().get("from").get("id").asText()) : cloudEvent.getEntityId();
+
+        TbProtoQueueMsg<ToCloudEventMsg> cloudEventMsg = new TbProtoQueueMsg<>(entityId, toCloudEventMsg);
+
+        producer.send(tpi, cloudEventMsg, null);
+    }
+
+    private TbQueueProducer<TbProtoQueueMsg<ToCloudEventMsg>> chooseProducer(boolean isTS) {
+        return isTS ? tbCloudEventProvider.getCloudEventTSMsgProducer() : tbCloudEventProvider.getCloudEventMsgProducer();
+    }
+
+    @Override
+    public PageData<CloudEvent> findCloudEvents(TenantId tenantId, Long seqIdStart, Long seqIdEnd, TimePageLink pageLink) {
+        return createPageData(tenantId, false, cloudEventSettings.getPollInterval());
+    }
+
+    @Override
+    public PageData<CloudEvent> findTsKvCloudEvents(TenantId tenantId, Long seqIdStart, Long seqIdEnd, TimePageLink pageLink) {
+        return createPageData(tenantId, true, cloudEventTSSettings.getPollInterval());
+    }
+
+    @NotNull
+    private PageData<CloudEvent> createPageData(TenantId tenantId, boolean isTS, long pollInterval) {
+        TbQueueConsumer<TbProtoQueueMsg<ToCloudEventMsg>> consumer = chooseConsumer(isTS);
+        TopicPartitionInfo tpi = new TopicPartitionInfo(consumer.getTopic(), tenantId, 1, true);
+
+        subscribe(consumer, tpi);
+
+        List<TbProtoQueueMsg<ToCloudEventMsg>> cloudMessages = consumer.poll(pollInterval);
+        List<CloudEvent> cloudEvents =
+                cloudMessages.stream()
+                        .map(msg -> ProtoUtils.fromProto(msg.getValue().getCloudEventMsg()))
+                        .toList();
+
+        return new PageData<>(cloudEvents, 0, cloudEvents.size(), false);
+    }
+
+    private void subscribe(TbQueueConsumer<TbProtoQueueMsg<ToCloudEventMsg>> consumer, TopicPartitionInfo tpi) {
+        if (!consumer.getFullTopicNames().contains(tpi.getFullTopicName())) {
+            consumer.subscribe(Collections.singleton(tpi));
+        }
+    }
+
+    @Override
+    public void unsubscribeConsumers() {
+        tbCloudEventProvider.getCloudEventMsgConsumer().unsubscribe();
+        tbCloudEventProvider.getCloudEventTSMsgConsumer().unsubscribe();
+    }
+
+    @Override
+    public void commit(boolean isTS) {
+        chooseConsumer(isTS).commit();
+    }
+
+    private TbQueueConsumer<TbProtoQueueMsg<ToCloudEventMsg>> chooseConsumer(boolean isTS) {
+        return isTS ? tbCloudEventProvider.getCloudEventTSMsgConsumer() : tbCloudEventProvider.getCloudEventMsgConsumer();
+    }
+
+    @Override
+    public void cleanupEvents(long ttl) {
+        throw new UnsupportedOperationException(METHOD_CANNOT_BE_USED_FOR_THIS_SERVICE);
```

Comment:
If something is stored in DB, let's leave this method with cleanup() using EdgeEventDao bean or let's use CloudEventDao inside CloudEventsCleanUpService

Reply:
It's a good idea:)

Reply:
What about this?

Reply:
DONE)